src/allocated_nonnative_field_var.rs

```diff
@@ -530,6 +530,86 @@ impl<TargetField: PrimeField, BaseField: PrimeField>
             OptimizationGoal::Weight => OptimizationType::Weight,
         }
     }
+
+    /// Allocates a new non-native field variable with value given by the function `f`.  If `mode`
+    /// is [AllocationMode::Witness], then this function checks that the field element has value in
+    /// `[0, modulus)`, and returns the bits of its binary representation. The bits are in
+    /// little-endian (i.e., the bit at index 0 is the LSB) and the bit-vector is empty in
+    /// non-witness allocation modes.
+    pub fn new_variable_alloc_le_bits<T: Borrow<TargetField>>(
+        cs: impl Into<Namespace<BaseField>>,
+        f: impl FnOnce() -> Result<T, SynthesisError>,
+        mode: AllocationMode,
+    ) -> R1CSResult<(Self, Vec<Boolean<BaseField>>)> {
+        let ns = cs.into();
+        let cs = ns.cs();
+
+        let optimization_type = match cs.optimization_goal() {
+            OptimizationGoal::None => OptimizationType::Constraints,
+            OptimizationGoal::Constraints => OptimizationType::Constraints,
+            OptimizationGoal::Weight => OptimizationType::Weight,
+        };
+
+        let params = get_params(
+            TargetField::size_in_bits(),
+            BaseField::size_in_bits(),
+            optimization_type,
+        );
+        let zero = TargetField::zero();
+
+        let elem = match f() {
+            Ok(t) => *(t.borrow()),
+            Err(_) => zero,
+        };
+        let elem_representations = Self::get_limbs_representations(&elem, optimization_type)?;
+        let mut limbs = Vec::new();
+
+        for limb in elem_representations.iter() {
+            limbs.push(FpVar::<BaseField>::new_variable(
+                ark_relations::ns!(cs, "alloc"),
+                || Ok(limb),
+                mode,
+            )?);
+        }
+
+        let num_of_additions_over_normal_form = if mode != AllocationMode::Witness {
+            BaseField::zero()
+        } else {
+            BaseField::one()
+        };
+
+        let mut bits = Vec::new();
+
+        if mode == AllocationMode::Witness {
+            for limb in limbs.iter().rev().take(params.num_limbs - 1) {
+                bits.extend(
+                    Reducer::<TargetField, BaseField>::limb_to_bits(limb, params.bits_per_limb)?
+                        .into_iter()
+                        .rev(),
+                );
+            }
+
+            bits.extend(
+                Reducer::<TargetField, BaseField>::limb_to_bits(
+                    &limbs[0],
+                    TargetField::size_in_bits() - (params.num_limbs - 1) * params.bits_per_limb,
+                )?
+                .into_iter()
+                .rev(),
+            );
+        }
```

Comment:
I think when the mode is not `Witness`, we can still generate bits easily:

* if `mode == Constant`, we can return `Boolean::Constant` values (this requires no constraints)
* if `mode == Input`, we can just reuse the existing machinery for the `mode == Witness` case.

If the concern is adding potentially unnecessary constraints, we can make a helper function which generates bits only if a flag is true.

Reply:
Yes, my concern is that when `mode == Input`, our current allocation gadget, which uses this function as a black box, would generate the bits, which would be unnecessary.

Maybe let us extend this function so that one can specify, via a bool, whether or not bits are wanted.